src/binxml/assemble.rs

```diff
@@ -282,119 +265,120 @@ pub fn create_record_model<'a>(
+            for token in template_def.tokens {
+                if let BinXMLDeserializedTokens::Substitution(ref substitution_descriptor) = token {
+                    expand_token_substitution(&mut template, substitution_descriptor, chunk, stack)
+                } else {
+                    _expand_templates(Cow::Owned(token), chunk, stack);
+                }
             }
+        }
+        return;
+    };
+
+    for token in template_def.tokens.iter() {
+        if let BinXMLDeserializedTokens::Substitution(ref substitution_descriptor) = token {
+            expand_token_substitution(&mut template, substitution_descriptor, chunk, stack)
         } else {
-            _expand_templates(Cow::Borrowed(token), stack);
+            _expand_templates(Cow::Borrowed(token), chunk, stack);
         }
     }
 }
 
 fn _expand_templates<'a>(
     token: Cow<'a, BinXMLDeserializedTokens<'a>>,
+    chunk: &'a EvtxChunk<'a>,
     stack: &mut Vec<Cow<'a, BinXMLDeserializedTokens<'a>>>,
 ) {
     match token {
         // Owned values can be consumed when flatting, and passed on as owned.
-        Cow::Owned(BinXMLDeserializedTokens::Value(Cow::Owned(BinXmlValue::BinXmlType(
-            tokens,
-        )))) => {
+        Cow::Owned(BinXMLDeserializedTokens::Value(BinXmlValue::BinXmlType(
+                                                       tokens,
+                                                   ))) => {
             for token in tokens.into_iter() {
-                _expand_templates(Cow::Owned(token), stack);
+                _expand_templates(Cow::Owned(token), chunk, stack);
             }
         }
 
-        // All borrowed values are flattened and kept borrowed.
-        Cow::Owned(BinXMLDeserializedTokens::Value(Cow::Borrowed(BinXmlValue::BinXmlType(
-            tokens,
-        ))))
-        | Cow::Borrowed(BinXMLDeserializedTokens::Value(Cow::Owned(BinXmlValue::BinXmlType(
-            tokens,
-        ))))
-        | Cow::Borrowed(BinXMLDeserializedTokens::Value(Cow::Borrowed(BinXmlValue::BinXmlType(
-            tokens,
-        )))) => {
+        Cow::Borrowed(BinXMLDeserializedTokens::Value(BinXmlValue::BinXmlType(
+                                                          tokens,
+                                                      ))) => {
             for token in tokens.iter() {
-                _expand_templates(Cow::Borrowed(token), stack);
+                _expand_templates(Cow::Borrowed(token), chunk, stack);
             }
         }
 
         // Actual template handling.
         Cow::Owned(BinXMLDeserializedTokens::TemplateInstance(template)) => {
-            expand_owned_template(template, stack);
+            expand_template(template, chunk, stack);
         }
         Cow::Borrowed(BinXMLDeserializedTokens::TemplateInstance(template)) => {
-            expand_borrowed_template(template, stack);
+            // We never actually see this in practice, so we don't mind paying for `clone` here.
```

Comment:
We need to mention when this happens - what is the flow that leads to having a borrowed template in this state?
I don't think we could reason about what `We never actually see this in practice` in the future.